### PR TITLE
Back point selection with a psygnal Selection

### LINKED ISSUE #5686: feature request: point selection event
## 🚀 Feature
I would like an event to be fired when the selection changes on the points layer

## Motivation

I would like to use the point selection as the model and single source of truth for selection in a plugin. There is no event on selection change.

## Pitch

Add an event for selection change

## Alternatives

there is a 'highlight' event but this is fired when hovering over a point too

### PR DESCRIPTION
**Pull request by katherine-hutchings**
# Description
This pull request makes the point selection model use an evented set from psygnal rather than the python set. This will close #5686 and also #4023. 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# References
https://psygnal.readthedocs.io/en/latest/API/containers/#psygnal.containers.EventedSet

# How has this been tested?
Manually tested, but have not yet added tests to the test suite.
@alisterburt helped me with this PR and will add some tests. 

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by alisterburt** [comment]
great work @katherine-hutchings - this will be really useful to have!

should also facilitate #5030 

@kevinyamauchi or @jni would you mind giving a quick review here? Is a super small change

**Review by alisterburt — APPROVED** [review]
approving but someone else should also approve before merge as I helped here

**Review by brisvag — COMMENTED** [review]
In principle looks good as is (and it's fine to merge this as an incremental change) but I just wanted to mention that we have the `Selection` model which might be a better fit here. There's a very old PR where @jni and I were trying to do this for the Labels layer, which might be a good starting point if you wanted to push this a bit further: https://github.com/napari/napari/pull/2978.

Even if not, it might be worth checking the name convention of `Selection` events and reflect it here, just so we don't have to deprecate stuff in the future when/if we do the swap :)

**Review by brisvag — APPROVED** [review]

**Review by alisterburt — CHANGES_REQUESTED** [review]
@brisvag thanks for the reveiw, have updated to the psygnal selection model - test failures are real so blocking merge here whilst I investigate

**Review by alisterburt — APPROVED** [review]

**Comment by alisterburt** [comment]
The bug was an interesting one, summarising below 

As we are now maintaining an evented set as the model of the selection we no longer replace the existing set with the incoming data in the `Points.selected_data` setter, we instead clear the existing set then update it with the new contents.

The `_toggle_selected` utility modified the set in-place, so clearing the existing set in the setter also modified the incoming set. I decided that the utility function should instead return a copy, this seems to be how it was intended based on usage `layer.selected_data = _toggle_selected(selection, value)`

**Comment by alisterburt** [comment]
test failures look unrelated, rerunning but I think this is good to go

**Comment by Czaki** [comment]
Test failure is connected with new PySide6 release and lack of bugfix in the app model. 
As bugfix is merged then @tlambert03 maybe it is time for 0.1.3 release?

**Comment by jni** [comment]
Hi @katherine-hutchings and welcome! 😊 

I'm A-ok with this change as it makes things strictly better, but @alisterburt what do you think about @brisvag's comment about our existing Selection object? I'd forgotten about it... 😅 Having said that, my main question is which API do we think is the future of the library that we will harmonise all selections to? If it's psygnal then we don't have to worry about it. But if it's Selection, or some more complex wrapper of the psygnal set, then we might want to wait a hot minute and consider it?

**Comment by alisterburt** [comment]
thanks for the review @jni !

We originally had a bare `EventedSet` here which I switched over to the `Selection` from psygnal after @brisvag suggested the selection model - my vote is for harmonising around psygnal in napari generally and selections seem like a good place to start but happy to swap over to our internal vispy backed `Selection` if we're worried about mixing event systems

I was planning a followup PR which unifies layers which support selection around `Layer.selection` rather than the mishmash we have currently of `selected_data` `selected_label` - could be a good moment to also unify the object backing the selections?

**Comment by brisvag** [comment]
> I was planning a followup PR which unifies layers which support selection around `Layer.selection` rather than the mishmash we have currently of `selected_data` `selected_label` - could be a good moment to also unify the object backing the selections?

Awesome :)

**Review by kevinyamauchi — APPROVED** [review]
This looks good to me. Thanks @katherine-hutchings ! This will definitely make my life easier.

Given that we have two approvals and confirmation from @Czaki  that the remaining test failure is unrelated, I think we should merge soon.

> I was planning a followup PR which unifies layers which support selection around Layer.selection rather than the mishmash we have currently of selected_data selected_label - could be a good moment to also unify the object backing the selections?

This would be great, @alisterburt !

**Comment by alisterburt** [comment]
Thanks all - I'm going to assume the thumbs up from jni is a vote of approval, will merge tomorrow morning

**Comment by andy-sweet** [comment]
I marked this with the 0.4.18 milestone because this is a small, non-breaking (because `Selection` is a `MutableSet`) feature addition. Whoever manages the 0.4.18 release should feel free to bump this to 0.5.0 if desired.

**Pull request merged**
